song_player/SongPlayer.cpp

```diff
@@ -0,0 +1,209 @@
+#include "SongPlayer.h"
+
+namespace
+{
+const int MS_TO_MINUTES = 60000;
+const double MS_TO_SECONDS = 1000.0;
+const int PAGE_STEP_INCREMENTS = 10;
+}
+
+SongPlayer::SongPlayer(QWidget *parent) : QWidget(parent)
+{
+    createWidgets();
+    createShortcuts();
+
+    connect(&mediaPlayer, &QMediaPlayer::positionChanged, this, &SongPlayer::updatePosition);
+    connect(&mediaPlayer, &QMediaPlayer::durationChanged, this, &SongPlayer::updateDuration);
+    connect(&mediaPlayer, &QMediaObject::metaDataAvailableChanged, this, &SongPlayer::updateInfo);
+
+}
+
+void SongPlayer::openFile()
+{
+    const QStringList musicPaths = QStandardPaths::standardLocations(QStandardPaths::MusicLocation);
+    const QString filePath =
+        QFileDialog::getOpenFileName(this, tr("open file"),
```

Comment:
This entire function is re written in the Song Controller branch and contains fewer lines